jmclient/jmclient/yieldgenerator.py

```diff
@@ -112,6 +114,58 @@ def create_my_orders(self):
 
         return [order]
 
+    def get_fidelity_bond_template(self):
+        if not isinstance(self.wallet_service.wallet, FidelityBondMixin):
+            jlog.info("Not a fidelity bond wallet, not announcing fidelity bond")
+            return None
+        blocks = jm_single().bc_interface.get_current_block_height()
+        mediantime = jm_single().bc_interface.get_best_block_median_time()
+
+        BLOCK_COUNT_SAFETY = 2 #use this safety number to reduce chances of the proof expiring
+                               #before the taker gets a chance to verify it
+        RETARGET_INTERVAL = 2016
+        CERT_MAX_VALIDITY_TIME = 1
+        cert_expiry = ((blocks + BLOCK_COUNT_SAFETY) // RETARGET_INTERVAL) + CERT_MAX_VALIDITY_TIME
+
+        utxos = self.wallet_service.wallet.get_utxos_by_mixdepth(include_disabled=True,
+            includeheight=True)[FidelityBondMixin.FIDELITY_BOND_MIXDEPTH]
+        timelocked_utxos = [(outpoint, info) for outpoint, info in utxos.items()
+            if FidelityBondMixin.is_timelocked_path(info["path"])]
+        if len(timelocked_utxos) == 0:
+            jlog.info("No timelocked coins in wallet, not announcing fidelity bond")
+            return
+        timelocked_utxos_with_confirmation_time = [(outpoint, info,
+            jm_single().bc_interface.get_block_time(
+                jm_single().bc_interface.get_block_hash(info["height"])
+            ))
+            for (outpoint, info) in timelocked_utxos]
+
+        interest_rate = get_interest_rate()
+        max_valued_bond = max(timelocked_utxos_with_confirmation_time, key=lambda x:
+            FidelityBondMixin.calculate_timelocked_fidelity_bond_value(x[1]["value"], x[2],
+                x[1]["path"][-1], mediantime, interest_rate)
+        )
+        (utxo_priv, locktime), engine = self.wallet_service.wallet._get_key_from_path(
+            max_valued_bond[1]["path"])
+        utxo_pub = engine.privkey_to_pubkey(utxo_priv)
+        cert_priv = os.urandom(32) + b"\x01"
+        cert_pub = btc.privkey_to_pubkey(cert_priv)
+        cert_msg = b"fidelity-bond-cert|" + cert_pub + b"|" + str(cert_expiry).encode("ascii")
+        cert_sig = base64.b64decode(btc.ecdsa_sign(cert_msg, utxo_priv))
```

Comment:
Is the use of a separate certificate keypair documented? I would like to think about how this works in more detail.

Reply:
Documented now in my modifications to `Joinmarket-messaging-protocol.md `

Reply:
Yes this covers the details I was looking for, thanks.